statistics/handle/handle_hist.go

```diff
@@ -0,0 +1,375 @@
+}
+
+// AppendNeededColumn appends needed column to ch, if exists, do not append the duplicated one.
+func (h *Handle) AppendNeededColumn(c model.TableColumnID, resultCh chan model.TableColumnID, timeout time.Duration) error {
+	toTimout := time.Now().Local().Add(timeout)
+	colTask := &NeededColumnTask{TableColumnID: c, ToTimeout: toTimout, ResultCh: resultCh}
+	return h.writeToChanWithTimeout(h.StatsLoad.NeededColumnsCh, colTask, timeout)
+}
+
+var errExit = errors.New("Stop loading since domain is closed")
+
+type statsReaderContext struct {
+	reader      *statsReader
+	createdTime time.Time
+}
+
+// SubLoadWorker loads hist data for each column
+func (h *Handle) SubLoadWorker(ctx sessionctx.Context, exit chan struct{}, exitWg *sync.WaitGroup) {
+	readerCtx := &statsReaderContext{}
+	defer func() {
+		exitWg.Done()
+		logutil.BgLogger().Info("SubLoadWorker exited.")
+		if readerCtx.reader != nil {
+			err := h.releaseStatsReader(readerCtx.reader, ctx.(sqlexec.RestrictedSQLExecutor))
+			if err != nil {
+				logutil.BgLogger().Error("Fail to release stats loader: ", zap.Error(err))
+			}
+		}
+	}()
+	for {
+		err := h.handleOneTask(readerCtx, ctx.(sqlexec.RestrictedSQLExecutor), exit)
+		if err != nil {
+			switch err {
+			case errExit:
+				return
+			default:
+				time.Sleep(10 * time.Millisecond)
+				continue
+			}
+		}
+	}
+}
+
+// handleOneTask handles one column task.
+func (h *Handle) handleOneTask(readerCtx *statsReaderContext, ctx sqlexec.RestrictedSQLExecutor, exit chan struct{}) (err error) {
+	defer func() {
+		// recover for each task, worker keeps working
+		if r := recover(); r != nil {
+			buf := make([]byte, 4096)
+			stackSize := runtime.Stack(buf, false)
+			buf = buf[:stackSize]
+			logutil.BgLogger().Error("stats loading panicked", zap.String("stack", string(buf)))
+		}
+	}()
+	h.getFreshStatsReader(readerCtx, ctx)
+	task, err := h.drainColTask(exit)
+	if err != nil {
+		if err != errExit {
+			logutil.BgLogger().Error("Fail to drain task for stats loading.", zap.Error(err))
+		}
+		return err
+	}
+	col := task.TableColumnID
+	// to avoid duplicated handling in concurrent scenario
+	if !h.setWorking(col, task.ResultCh) {
+		return nil
+	}
+	oldCache := h.statsCache.Load().(statsCache)
+	tbl, ok := oldCache.tables[col.TableID]
+	if !ok {
+		task.ResultCh <- col
+		return nil
+	}
+	c, ok := tbl.Columns[col.ColumnID]
+	if !ok || c.Len() > 0 {
+		task.ResultCh <- col
+		return nil
+	}
+	t := time.Now()
+	hist, err := h.readStatsForOne(col, c, readerCtx.reader)
+	if err != nil {
+		h.StatsLoad.NeededColumnsCh <- task
+		return err
+	}
+	metrics.ReadStatsHistogram.Observe(float64(time.Since(t).Milliseconds()))
+	if hist != nil && h.updateCachedColumn(col, hist) {
+		task.ResultCh <- col
+	}
+	h.finishWorking(col)
+	return nil
+}
+
+func (h *Handle) getFreshStatsReader(readerCtx *statsReaderContext, ctx sqlexec.RestrictedSQLExecutor) {
+	if readerCtx.reader == nil || readerCtx.createdTime.Add(h.Lease()).Before(time.Now()) {
```

Comment:
Why we need to check `createdTime`? Is it because we don't want to make the txn run too long?

Reply:
yes.